### ISSUE TITLE
[Feature Request] Distributed Tensorflow with Data Parallelism and DMA

### ISSUE THREAD

**Issue by jongsae** [tensorflow/tensorflow#11196]
Hi, I have a machine with multiple GPUs and deployed a distributed Tensorflow exploiting data parallelism. I use ClusterConfig to configure the cluster's topology and pass this to Experiment to run the distributed training. I wanted to make one of the GPUs as the parameter server and use the rest for the workers, each of which uses one GPU. I did this through launching multiple processes for parameter server and workers while setting up the env variable (CUDA_VISIBLE_DEVICES) to one of the GPUs. 

In this case, I noticed that the GPUs don't communicate to each other via DMA and I guess this is because the DMA is only available between visible devices to Tensorflow.  

I also tried "device_count={"GPU": 1}" to make Tensorflow see all the GPUs while using only one GPU, but this still seems to occupy the whole resource of all visible GPUs, preventing another Tensorflow process to be launched over the idle GPU. 

It would be great if there is a way to use only one GPU for a Tensorflow process but still enable DMA with the other GPUs sitting in the system. Am I missing such feature even though it exists? 

Thank you!

**Comment by byronyi** [comment]
Currently distributed TensorFlow is relatively not performant compared to single-process multi-device setup. If you are not running your job over multiple machines, you could consider using a intra process parameter server (which typically only uses CPU and host RAM as there is not much work to do for PS).

Nevertheless, we are preparing a patch on interprocess device DMA (GPU Direct) for distributed runtime. But currently our design only consider the use case that cross the physical machine boundary through an RDMA-capable network interface card (rNIC). Could you explain a little bit more about your use case? Why do you want to run distributed TF on a single node? I am assuming it's not for testing as you mentioned DMA and thus only a performance-wise matter.

If your use case is valid, we might consider adding a shared-memory based transport. This may not be as efficient as GPU p2p (I don't know how to do that across process boundary), but certainly faster than current gRPC runtime as it doesn't go through the socket.

**Comment by jongsae** [comment]
Hi Bairen, first, thanks for your response! 

The reason why I wanted to deploy a distributed TF on a single node is that my application is not a very good fit for exploiting model parallelism. However, I have multi GPUs available in a machine so I would like to exploit data parallelism. 

I agree with you that having an intra process parameter server could be the solution for my case. However, the problem here is the fact that I would like to have a distributed version of the pre-existing implementation that uses Tensorflow's [Experiment](https://www.tensorflow.org/api_docs/python/tf/contrib/learn/Experiment). As far as I understand, the only way to parallelize the Experiment-based application for data parallelism is to use ClusterConfig, which requires to launch multiple processes. 

Re-developing the application from scratch to enable the intra process parameter server imposes a lot of engineering effort on me and I believe other developers could have similar use-cases like this if they don't develop TF application from scratch. 

Then your question could be why I need DMA. My application has a large size of model parameters that need to be communicated and the CPU parameter server requires the communication to go through PCIe, which is not very performant. Therefore, I would like to deploy my parameter server on one of the GPUs and launch the workers on the other GPUs and have them to communicate via DMA for performant communication. (I would be even happier if Tensorflow allows me to place a parameter server with a worker on a GPU since then I could minimize the waste of compute resource. I am not quite sure if this is possible in the current implementation.)

Although I have found people who already use a single machine for distributed TF, I think my case will be more common as the number of GPUs per machine increases because it would be unlikely for people to always assign all GPUs available in a machine to a single model replica. Moreover, as the learning model gets larger (e.g., RNNs), the PCIe communication of CPU parameter server will eventually be impractical. 

I think the best way of handling this issue is that TF launches a single process for parameter servers and workers sitting in the same machine and only communicate through gRPC when there is a need for machine-to-machine communication. The current way of having distributed TF through multi processes with CUDA_VISIBLE_DEVICES effectively disables DMA, which limits the use case like mine.

Please feel free to correct me if any of my understanding is wrong. Thank you!

**Comment by byronyi** [comment]
@martinwicke Any comments?

**Comment by byronyi** [comment]
Seems the [`cudaIpc*` functions](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g02bb3632b5d223db6acae5f8744e2c91) could serve as a loopback transport for interprocess p2p. I have not used these functions so I am not sure.

**Comment by jongsae** [comment]
The cudaIPc* functions indeed seem to be the hope for interprocess p2p communication. I am not sure how the CUDA_VISIBLE_DEVICES play in this scenario though since the processes see different sets of GPUs (e.g., PS process sees GPU#0 while Worker process sees GPU#1). 

I wondered if you would consider incorporating this feature to the next patch? In any case, I really appreciate your effort!

**Comment by cy89** [comment]
@poxvoculi can you shed any light on this discussion?

**Comment by poxvoculi** [comment]
The analysis of the situation by @jongsae and @byronyi seems largely correct.  Inter-process communication is less performant than intra-process, so the later is to be preferred when using multiple GPUs on a single machine.   Separate processes cannot access each other's address space, so TF will always try to use an RPC when communicating between processes, even if they're on the same machine.  Although DMA will be used in moving data on/off GPUs, there will be lots of extra useless work to execute the RPC whose data packets don't actually exit the local machine.

My first thought would be to restructure the program construction so that the elements which are now executed as independent processes could be packaged either as subgraphs assigned to a single worker, or subgraphs assigned to different workers, depending on whether real (cross-process) distribution is actually required.  I understand this may be difficult when trying to reuse software not designed for that purpose.  Perhaps @saeta  has some useful observations on the general problem of deploying data parallel programs with one versus multiple processes.

**Comment by jongsae** [comment]
@poxvoculi Thanks for your answer! Following your suggestion, I will consider re-implementing the existing TF application such that the parameter server and workers sit in the same process and they use the intra-process communication. @saeta, I am looking forward to hearing regarding your experience in deploying data parallel programs in a single process. 

In a short-term perspective, I agree with this approach. However, in a long run,  as I argued earlier in my second comment in this issue, I believe it would be useful for many use-cases that TF incorporates the inter-process DtoD communication to support multi-process data parallel TF programs. An alternative approach would be to modify TF wrappers (e.g., Estimator or Experiment) so that they support single-process distributed TF in case where the configured parameter servers and workers (e.g., via tf.train.ClusterSpec or tf.contrib.learn.python.learn.estimators.run_config) are placed in the same local machine, which I think, is a better (or cleaner) approach. 

I wondered what TF folks @byronyi @martinwicke @cy89 @poxvoculi think about it?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
We are implementing better support for multi-GPU systems (available in `tf.contrib.estimator`), which will not rely on multiple processes. 

I will close this issue, it is too diffuse as a feature request. Please reopen a new issue with concrete suggestions.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.
